api/service.proto

```diff
@@ -0,0 +1,103 @@
+syntax = "proto3";
+
+import "github.com/kubernetes-csi/csi-proxy/api/file_system.proto";
+import "github.com/kubernetes-csi/csi-proxy/api/general_storage.proto";
+import "github.com/kubernetes-csi/csi-proxy/api/iscsi.proto";
+import "github.com/kubernetes-csi/csi-proxy/api/smb.proto";
+
+service CSIProxyService {
+    // ********************** //
+    // File system operations //
+    // ********************** //
+
+    // PathExists checks if the given path exists on the host.
+    rpc PathExists(PathExistsRequest) returns (PathExistsResponse) {}
+
+    // Mkdir creates a directory at the requested path on the host.
+    rpc Mkdir(MkdirRequest) returns (MkdirResponse) {}
+
+    // Rmdir removes the directory at the requested path on the host.
+    // This may be used for unlinking a symlink created through LinkVolume.
+    rpc Rmdir(RmdirRequest) returns (RmdirResponse) {}
+
+    // LinkPath invokes mklink on the global staging path of a volume linking it to a path within a container.
+    rpc LinkPath(LinkPathRequest) returns (LinkPathResponse) {}
+
+    // ********************************* //
+    // General storage/volume operations //
+    // ********************************* //
+
+    // Rescan refreshes the host storage cache.
+    rpc Rescan(RescanRequest) returns (RescanResponse) {}
+
+    // PartitionDisk initializes and partitions a disk device (if the disk has not
+    // been partitioned already) and returns the resulting volume object's GUID.
+    rpc PartitionDisk(PartitionDiskRequest) returns (PartitionDiskResponse) {}
+
+    // FormatVolume formats a volume with the provided file system.
+    // The resulting volume is mounted at the requested global staging path.
+    rpc FormatVolume(FormatVolumeRequest) returns (FormatVolumeResponse) {}
+
+    // ListDiskLocations returns locations <Adapter, Bus, Target, LUN ID> of all disk devices enumerated by Windows.
+    rpc ListDiskLocations(ListDiskLocationsRequest) returns (ListDiskLocationsResponse) {}
+
+    // ListDiskIDs returns all IDs (from IOCTL_STORAGE_QUERY_PROPERTY) of all disk devices enumerated by Windows.
+    rpc ListDiskIDs(ListDiskIDsRequest) returns (ListDiskIDsResponse) {}
+
+    // ListDiskVolumeMappings returns a map of all disk devices and volumes GUIDs
+    rpc ListDiskVolumeMappings(ListDiskVolumeMappingsRequest) returns (ListDiskVolumeMappingsResponse) {}
```

Comment:
What are the use cases of these three list functions? Is there some doc or link you can share?

Reply:
`ListDiskLocations` or `ListDiskIDs` will be used by a block based CSI node driver to identify a disk device published to a node. 

`ListDiskIDs` maps to invoking `scsi-id` in Linux as done at https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gcepd/gce_util.go#L231 as well as what `Get-GcePdName` achieves internally when invoked through: https://github.com/kubernetes/kubernetes/blob/17997ab7161736b23359480c774415bda3877130/pkg/volume/gcepd/attacher.go#L180. All SCSI Page 83 IDs (along with other IDs) for all disk devices will be returned here.

`ListDiskLocations` maps to invoking the `Get-Disk` cmdlet and looking at standard disk fields as done at https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/azure_common_windows.go#L42

`ListDiskVolumeMappings` is required to obtain a mapping of volume device GUIDs as enumerated by Windows to the disk device objects backing the volumes. This information will be used by a block based CSI driver to enumerate the volume devices hosted on a disk (published by the CSI attacher). The volume device GUID for a volume can be passed to `MountVolume` to mount the volume at the desired global staging path.

Reply:
Adding comments